Autopilot/SensorFusion/Src/SensorFusion.cpp

```diff
@@ -0,0 +1,445 @@
+* Madgwick is used for estimating attitude.
+* A custom Kalmann filter is used along with the estimated attitude to estimate position.
+* Author: Lucy Gong, Dhruv Rawat, Anthony Bertnyk
+*/
+#include "SensorFusion.hpp"
+#include "MadgwickAHRS.h"
+#include <math.h>
+#include "imu.hpp"
+#include "gps.hpp"
+#include "altimeter.hpp"
+#include "airspeed.hpp"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+    #include "CControlFunctions.h"
+#ifdef __cplusplus
+}
+#endif
+
+typedef struct {
+    float roll, pitch, yaw; //rad
+    float rollRate, pitchRate, yawRate; //rad/s
+    float airspeed; //m/s
+} SFAttitudeOutput_t ;
+
+struct SFPositionOutput_t{
+    float altitude; //m
+    float rateOfClimb; //m/s
+    long double latitude; //Decimal degrees
+    float latitudeSpeed; //m/s
+    long double longitude; //Decimal degrees
+    float longitudeSpeed; //m/s
+};
+
+const int NUM_KALMAN_VALUES = 6; 
+
+struct SFIterationData_t{
+    float prevX[NUM_KALMAN_VALUES];
+    float prevP[NUM_KALMAN_VALUES*NUM_KALMAN_VALUES];
+};
+
+static IMU *imuObj;
+static Gps *gpsObj;
+static Altimeter *altimeterObj;
+static airspeed *airspeedObj;
+static SFIterationData_t iterData;
+static SFOutput_t SFOutput;
+
+void SF_Init(void)
+{
+#ifdef TARGET_BUILD
+
+  imuObj = &BMX160::getInstance();
+  gpsObj = NEOM8::GetInstance();
+  //Waiting for definitions
+  //altimeterObj = MS5637::GetInstance();
+  //airspeedObj = dummyairspeed::GetInstance();
+
+#elif defined(UNIT_TESTING)
+  imuObj = TestIMU::GetInstance();
+  gpsObj = TestGps::GetInstance();
+  //altimeterObj = TestAltimeter::GetInstance();
+  //airspeedObj = TestAirspeed::GetInstance();
+#endif
+}
+
+void localToGlobalAccel(IMUData_t *imudata, float *u)
+{
+    u[0] = imudata->accy; //Vertical acceleration
+    u[1] = imudata->accx; //Latitudinal acceleration
+    u[2] = imudata->accz; //Longitudinal acceleration
+}
+
+SFError_t SF_GetAttitude(SFAttitudeOutput_t *Output, IMUData_t *imudata, airspeedData_t *airspeeddata) {
+    
+    //Error output
+    SFError_t SFError;
+
+    SFError.errorCode = 0;
+
+    //IMU integration outputs
+    float imu_RollAngle = 0;
+    float imu_PitchAngle = 0;
+    float imu_YawAngle = 0;
+
+    float imu_RollRate = 0;
+    float imu_PitchRate = 0;
+    float imu_YawRate = 0;
+    
+    //Airspeed checks are temporarily disabled until an airspeed driver is implemented
+
+    //Abort if both sensors are busy or failed data collection
+    if(imudata->sensorStatus != 0 ) // || airspeeddata->sensorStatus != 0)
+    {  
+
+        /************************************************************************************************
+         * THIS WILL PUT THE STATE MACHINE INTO FATAL FAILURE MODE... WE NEED TO DECIDE IF THIS IS WHAT
+         * WE WANT OR IF WE SHOULD RETHINK HOW WE WANT THIS MODULE TO RETURN A SENSOR ERROR! 
+         ************************************************************************************************/
+
+        SFError.errorCode = -1;
```

Comment:
This is a bit dangerous because if we return anything other than 0 the plane will switch over to safety... If you could come up with a list of error codes for sensor fusion and let us know which ones should cause fatal behaviour, that would be great.